### ISSUE TITLE
Light 2D with Shadows enabled does not work on Tile Set with Normal Map

### ISSUE THREAD

**Issue by DocoDan** [godotengine/godot#28129]
**Godot version:**

3.1.stable.mono.official

**OS/device including version:**

**Issue description:**

I am running into issues with the Light 2D in combination with a normal map.
I can either activate Shadows or the Normal Map but not both at the same time without Light-2D generating artifacts.
If I disable Shadows the simple Normal Map Tile Map in my Example-Project works fine.
If I however activate the Shadows on the Light-2D  and move it across the Tilemap with the Normal map TileSet, only the Tile which is the center of the Light 2D position and its occluder is lit. I would expect the normal map wall textures to be nicely lit while the tile is also able to cast shadows.
I have checked that I use the correct background color (RGB) for the normal map texture. The Light gradient uses no alpha and it makes no difference if I import the normal map texture as Normal Map "Detect" or just "Normal Map".
Please find attached the most simple version which uses two Tile Maps in the main World.scene - three unlit rows of tiles above and three normal map lines of tiles below. Shadows are enabled.

**Steps to reproduce:**
Move the Light 2D in the World.scene across the Tilemaps in the Editor in Editor Mode or while running the Scene. Enable and disable Shadows on the Light 2D.

**Minimal reproduction project:**

[TileMapNormalMapIssue.zip](https://github.com/godotengine/godot/files/3090123/TileMapNormalMapIssue.zip)

**Comment by TheDuriel** [comment]
Looks like everything is working as expected in your project.

Your occluder is drawing a shadow over the tile, so the light is no longer affecting it and you do not get normal map lighting.

By setting the occluders cull mode to clockwise, you will correctly see the effect of the normal map on your tiles. This will however require appropriate occluders to be set up.

Alternatively you can add a gradient of about 4-8 pixels to your light.

![image](https://user-images.githubusercontent.com/44248915/56336635-64c87b00-61a1-11e9-8210-b5d256715ee0.png)

**Comment by DocoDan** [comment]
Looks good to me in the screenshot you sent.
I did like you said by setting the occluder to clockwise and/or adding the
gradient.
Still I have the effect, that the cast shadow does not stop at the next
tile on the same level, but continues and also casts to the right endlessly
if there is a tile.
Do you know the reason? I would expect the shadow only to be seen if there
is no tile on the same level.
Am I a missing a point with appropriate occluder setup here?

[image: Shadows again.png]

**Comment by DocoDan** [comment]
Ok, think I got it now.
The whole thing was a bit complicated, because the issue is very specific, I guess.
I first had to realize, that the occluders cull mode has to be set for reach and every tile in the tileset.
Also I added the gradient as you said and added the normal map to the tile set and the tiles as well.
Also I did not set the Lights cull mask to the ground, so the ground only uses the normal map.
I documented my steps here, in case anybody experiences similar issues.

![ShadowIssue](https://user-images.githubusercontent.com/2548272/56868709-414dcd80-69f6-11e9-879a-9d6594c7ee3f.gif)
